### PR TITLE
MM-60978 - Message export: Improve xml fields; fix delete semantics

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- MM-60978: Add xml fields that are necessary for reconstructing when a message was created, updated, deleted, or edited
- MM-60543: Add a `messageId` field which allows us to track which message an edited or deleted message is referring to.
- MM-61168: Fix how deleted posts are exported. Before they were always being exported twice (first as a created post, then as a deleted post). Now we are only exporting the created post if the post was actually created during the present run (required adding a JobStartTime to the job which is invariant across multiple restarts).
- Also making sure the update semantics work the same way.
- Fixes to tests for new semantics, and new tests for all the new code and e2e tests for happy paths and edge cases (messages that are created in one batch or in one job and then edited/deleted in the next).

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-60978
- https://mattermost.atlassian.net/browse/MM-61168
- https://mattermost.atlassian.net/browse/MM-60543

#### Release Note
Will be included in the feature branch release note.
```release-note
NONE
```

### CODE REVIEW

**Review by agnivade — COMMENTED** [review]
Great work on writing thorough tests. Left some questions.

File: server/enterprise/message_export/actiance_export/actiance_export.go [hunk 1]
```diff
@@ -224,6 +234,63 @@ func ActianceExport(rctx request.CTX, p Params) (shared.RunExportResults, error)
...
+	}
+	// Post is newly created:
+	// *post.PostCreateAt == *post.PostUpdateAt && (post.PostDeleteAt == nil || *post.PostDeleteAt == 0)
+	// but also fallback to this in case there is missing data, which is better than not exporting anything.
+	results.CreatedPosts++
+	return createdPostToExportEntry(post), results
+}
+
+func isEditedOriginalMsg(post *model.MessageExport) bool {
+	return post.PostDeleteAt != nil && *post.PostDeleteAt > 0 && post.PostOriginalId != nil && *post.PostOriginalId != ""
```
[the remark this answers is not in the document]
Comment (cpoile) [reply server/enterprise/message_export/actiance_export/actiance_export.go]: Thanks, done (for those parts where it works).

File: server/enterprise/message_export/actiance_export/actiance_export.go [hunk 2]
```diff
@@ -224,6 +234,63 @@ func ActianceExport(rctx request.CTX, p Params) (shared.RunExportResults, error)
...
+
+func isDeletedMsg(post *model.MessageExport) bool {
+	if post.PostDeleteAt != nil && *post.PostDeleteAt > 0 && post.PostProps != nil {
+		props := map[string]any{}
+		err := json.Unmarshal([]byte(*post.PostProps), &props)
+		if err != nil {
+			return false
+		}
+
+		if _, ok := props[model.PostPropsDeleteBy]; ok {
```
[the remark this answers is not in the document]
Comment (cpoile) [reply server/enterprise/message_export/actiance_export/actiance_export.go:288]: The `model.PostPropsDeleteBy` post prop? In `SqlPostStore.Delete`.

File: server/enterprise/message_export/actiance_export/actiance_export.go [hunk 3]
```diff
@@ -75,12 +75,32 @@ type LeaveExport struct {
...
-	Message      string   `xml:"Content"`      // the text body of the post
-	PreviewsPost string   `xml:"PreviewsPost"` // the post id of the post that is previewed by the permalink preview feature
+	XMLName   xml.Name `xml:"Message"`
+	MessageId string   `xml:"MessageId"`   // the message id in the db
+	UserEmail string   `xml:"LoginName"`   // the email of the person that sent the post
+	UserType  string   `xml:"UserType"`    // the type of the person that sent the post
+	CreateAt  int64    `xml:"DateTimeUTC"` // utc timestamp (unix milliseconds), the post's createAt
+
+	// Allows us to differentiate between:
+	// - "EditedOriginalMsg": the newly created message (new Id), which holds the pre-edited message contents. The
```
[the remark this answers is not in the document]
Comment (cpoile) [reply server/enterprise/message_export/actiance_export/actiance_export.go:85]: Yes, `EditedNewMsgId` will be blank for p2, because it is the EditedNewMsg.

File: server/enterprise/message_export/actiance_export/actiance_export.go [hunk 4]
```diff
@@ -224,6 +234,63 @@ func ActianceExport(rctx request.CTX, p Params) (shared.RunExportResults, error)
...
+	// - created:                         Id = new,  CreateAt = 1,    UpdateAt = 1, DeleteAt = 0
+	// - deleted:                         Id = orig, CreateAt = orig, UpdateAt = 2, DeleteAt = 2, props: deleteBy
+	// - edited: old post gets "created": Id = new,  CreateAt = 1,    UpdateAt = 2, DeleteAt = 2, originalId: orig
+	//           existing post modified:  Id = orig, CreateAt = 1,    UpdateAt = 2, DeleteAt = 0
+	//
+	// We also have other ways for a post to be updated:
+	//  - a root post in a thread is replied to, when a reply is edited, or (as of 10.2) when a reply is deleted
+
+	post := posts[i]
+	if isEditedOriginalMsg(post) {
```
[the remark this answers is not in the document]
Comment (cpoile) [reply server/enterprise/message_export/actiance_export/actiance_export.go:250]: I'm not 100% sure, since it's from before my time. But it seems like the `DeleteByProp` is for messages that are actually deleted. If there is no `DeleteByProp` then it's something else (it can identify if this is an edit history message, but I'm not sure if it helps identify another type of message). In the case of the `isEditedOriginalMsg` check, we don't even need to check the `DeleteByProp` because the `post.PostOriginalId` is enough to know it's an edited original msg.

File: server/enterprise/message_export/actiance_export/actiance_export.go [hunk 5]
```diff
@@ -234,13 +301,100 @@ func channelHasActivity(cmhs []*model.ChannelMemberHistoryResult, startTime int6
...
+func deletedPostToExportEntry(post *model.MessageExport, newMsg string) PostExport {
+	userType := "user"
+	if post.IsBot {
+		userType = "bot"
+	}
+	return PostExport{
+		MessageId:    *post.PostId,
+		CreateAt:     *post.PostCreateAt,
+		UpdateAt:     *post.PostDeleteAt,
+		UpdatedType:  shared.Deleted,
```
[the remark this answers is not in the document]
Comment (cpoile) [reply server/enterprise/message_export/actiance_export/actiance_export.go:329]: I definitely thought about combining these, but decided that being explicit what type of entry we are creating was far more important.

File: server/enterprise/message_export/actiance_export/actiance_export_test.go [hunk 6]
```diff
@@ -2013,3 +1642,130 @@ func Test_channelHasActivity(t *testing.T) {
...
+		}, originalPost)
+		require.NoError(t, err)
+		require.NotEqual(t, 0, originalPost.UpdateAt, "originalPost's updateAt was zero, test 2")
+		require.NotEqual(t, 0, post.UpdateAt, "edited post's updateAt was zero, test 2")
+		posts = append(posts, post)
+
+		simultaneous := post.UpdateAt
+
+		// Add 8 other posts at the same updateAt
+		for i := 1; i <= 8; i++ {
```
[the remark this answers is not in the document]
Comment (cpoile) [reply server/enterprise/message_export/actiance_export/actiance_export_test.go]: Hah, consequence of adding it after. Will switch those.

File: server/enterprise/message_export/actiance_export/actiance_export.go [hunk 7]
```diff
@@ -75,12 +75,32 @@ type LeaveExport struct {
 
 // The Message element indicates the message sent by a user
 type PostExport struct {
-	XMLName      xml.Name `xml:"Message"`
-	UserEmail    string   `xml:"LoginName"`    // the email of the person that sent the post
-	UserType     string   `xml:"UserType"`     // the type of the person that sent the post
-	PostTime     int64    `xml:"DateTimeUTC"`  // utc timestamp (seconds), time at which the user sent the post. Example: 1366611728
```
Comment (streamer45) [inline server/enterprise/message_export/actiance_export/actiance_export.go:81]: Is removing this going to cause issues in terms of compatibility? Asking as I am not sure how these exports are managed/parsed.

File: server/enterprise/message_export/actiance_export/actiance_export.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (streamer45) [reply server/enterprise/message_export/actiance_export/actiance_export.go:288]: Could this prop be tampered with by a malicious client?

File: server/enterprise/message_export/actiance_export/actiance_export.go [hunk 8]
```diff
@@ -235,13 +302,100 @@ func channelHasActivity(cmhs []*model.ChannelMemberHistoryResult, startTime int6
 	return false
 }
 
-func postToExportEntry(post *model.MessageExport, createTime *int64, message string) *PostExport {
+func createdPostToExportEntry(post *model.MessageExport) PostExport {
```
Comment (streamer45) [inline server/enterprise/message_export/actiance_export/actiance_export.go:305]: Here and in a few methods below I assume we are guaranteeing post can't be `nil`.

File: server/enterprise/message_export/actiance_export/actiance_export.go [hunk 9]
```diff
@@ -75,12 +75,32 @@ type LeaveExport struct {
...
-	XMLName      xml.Name `xml:"Message"`
-	UserEmail    string   `xml:"LoginName"`    // the email of the person that sent the post
-	UserType     string   `xml:"UserType"`     // the type of the person that sent the post
-	PostTime     int64    `xml:"DateTimeUTC"`  // utc timestamp (seconds), time at which the user sent the post. Example: 1366611728
-	Message      string   `xml:"Content"`      // the text body of the post
-	PreviewsPost string   `xml:"PreviewsPost"` // the post id of the post that is previewed by the permalink preview feature
+	XMLName   xml.Name `xml:"Message"`
+	MessageId string   `xml:"MessageId"`   // the message id in the db
+	UserEmail string   `xml:"LoginName"`   // the email of the person that sent the post
+	UserType  string   `xml:"UserType"`    // the type of the person that sent the post
```
Comment (streamer45) [inline server/enterprise/message_export/actiance_export/actiance_export.go:81]: Can we list the possible values? (i.e. `user`, `bot`).

File: server/enterprise/message_export/actiance_export/actiance_export.go [hunk 10]
```diff
@@ -235,13 +302,100 @@ func channelHasActivity(cmhs []*model.ChannelMemberHistoryResult, startTime int6
...
+		UserEmail:    *post.UserEmail,
+		PreviewsPost: post.PreviewID(),
+	}
+}
+
+func deletedPostToExportEntry(post *model.MessageExport, newMsg string) PostExport {
+	userType := "user"
+	if post.IsBot {
+		userType = "bot"
+	}
```
Comment (streamer45) [inline server/enterprise/message_export/actiance_export/actiance_export.go:324]: I am thinking a enum of strings could make sense here, so we get nice named constants.

File: server/enterprise/message_export/worker.go [hunk 11]
```diff
@@ -385,6 +398,13 @@ func extractJobData(logger *mlog.Logger, strmap map[string]string) (data shared.
 	// previous batches).
 	data.ExportPeriodStartTime = data.BatchStartTime
 
+	// JobStartTime is the start of this job. It is different from ExportPeriodStartTime because JobStartTime won't change
+	// if the job processes some batches, is stopped, and picked up again.
+	data.JobStartTime, err = strconv.ParseInt(strmap[JobDataStartTimestamp], 10, 64)
+	if err != nil {
+		return
+	}
```
Comment (streamer45) [inline server/enterprise/message_export/worker.go:406]: Unrelated but I am thinking that if any of these were to fail it would be quite hard to understand what exactly went wrong since we are not wrapping the error.

**Review by streamer45 — COMMENTED** [review]
Brilliant work, especially on ensuring this works, covering all sorts of edge cases :100: 

Left a couple of comments and questions, but nothing truly blocking.

**Review by agnivade — APPROVED** [review]

File: server/enterprise/message_export/actiance_export/actiance_export.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cpoile) [reply server/enterprise/message_export/actiance_export/actiance_export.go:288]: Yes, as discussed in our call. I'll make a post in the security channel with the findings from my investigation.

As for what would get exported, if there is no `deletedBy` prop, then this won't be considered a deleted post. If it was tampered with, then it will have an `updatedAt` > `createdAt` and so it will be considered an updated post (like a reaction). So we'll output it, but it won't be as a deleted post. 

I don't think we should make logic to deal with this because it's not something we should be expected to cover -- we would also have to start looking at all the combinations of post props and post fields and handle any combination that could be created through the api. And then maybe through store calls as well.

File: server/enterprise/message_export/actiance_export/actiance_export.go (hunk 7 above)
Comment (cpoile) [reply server/enterprise/message_export/actiance_export/actiance_export.go:81]: That might be an artifact of the github diff -- we're not removing any  of the existing xml fields, we're only adding new fields. There is one change to the meaning of one of the fields: `DateTimeUTC` used to mean different things; sometimes it was `createAt` other times it was `updateAt`/`deleteAt` depending on the post. Now it's always `createAt`. I'll be adding that to the doc notes and blog post.

Comment (streamer45) [reply server/enterprise/message_export/actiance_export/actiance_export.go:81]: `PostTime` field is gone as far as I can tell :thinking: . That's actually the line the comment referred to.

Comment (cpoile) [reply server/enterprise/message_export/actiance_export/actiance_export.go:81]: It's now the `CreateAt` field, which exports to the same  `xml:"DateTimeUTC"` field.

Comment (streamer45) [reply server/enterprise/message_export/actiance_export/actiance_export.go:81]: Oh, thanks, it's serialized to the same, totally missed that. Also it's in the milliseconds now right? A note sounds good, thanks :+1:

**Comment by cpoile** [comment]
Note: Flaky test is `api4.TestDeleteEmoji`, unrelated and fixed in master I think, so will merge.
